README.md
Goal:
1. Provide MVP implementation of new atomic swaps algorithm proposed in https://sites.cs.ucsb.edu/~victorzakhary/assets/papers/arXiv__Atomic_Cross_Chain_Commitment.pdf
2. Derisk implementation difficulty for full decentralized solution (port javascript over to Solidity)
3. Do all of the above without taking too much of Amr or Victor's time


To run:
1. Complete getting started tutorial on private Ethereum blockchain (i.e install Geth)
2. Start private ethereum blockchain by running
geth --datadir "./db2" --orkid 123456 --rpc --rpcport "8546"  --rpccorsdomain "*"  --port 30305 --nodiscover --rpcapi="admin,db,eth,debug,miner,net,shh,txpool,personal,web3" --ipcdisable console
2. In separate terminal, create new default account by running "geth account new --password <passwordfile>"
3. Unlock account by running "personal.unlockAccount(eth.accounts[0], 'yourpassword')" in terminal running private blockchain
4. If you want to try running verification from scratch, deploy "trent.sol" in Remix and set value == "1000" wei. Then copy contract address and change "smartContractAddress" queryBlockchain.js. Then run "node queryBlockchain.js"
5. If you want to just try what I've deployed, just run "node trent.js"
  
The WN contract is in "WN.sol"
The Trent smart contract is in "Trent.sol"
The blockchain query is in queryBlockchain.js and writes to folder /data
The verification algorithm is in trent.js and reads data from folder /data
I only use db2. I don't use db, but it can be used for testing

TODO: Copy my implementation in "trent.js" and "util.js" and take examples I've referenced (i.e. on-chain Merkle Proofs) below to finish final smart contract. The off-chain implementation of merkle proofs is in the folder /helpers


New Proposed Algorithm:
Alice deploys SC with validate WN smart contract (give base block of WN + block with WN smart contract that gets alice and bob’s agreement + blocks after)
Bob deploys SC with validate WN smart contract (give base block of WN + block with WN smart contract that gets alice and bob’s agreement + blocks after)
Bob and Alice can each provide the WN smart contract once the state has changed up starting from agreed upon base block AND smart contract depends on their addresses

Challenges: 
Had to understand how Ethereum internals truly worked (i.e. storage, merkle tries, address generation, etc.) and tried many workarounds not listed in research paper

Code Comment Guidelines:
1. Replace all comments that include "NOTE" with different hardcoded values.

Outside References for future on-chain verification:
1. https://github.com/aragon/evm-storage-proofs
2. https://github.com/ConsenSys/rb-relay/blob/master/contracts/MerklePatriciaProof.sol
3. https://github.com/zmitton/eth-proof
4. https://github.com/ethereum/btcrelay

Good Resources:
https://easythereentropy.wordpress.com/2014/06/04/understanding-the-ethereum-trie/
https://programtheblockchain.com/posts/2018/03/09/understanding-ethereum-smart-contract-storage/


db/history
tail -f myEth.log
person.newAccount("victor97")
eth.coinbase
eth.getBalance(eth.coinbase)
miner.start()
miner.stop()
eth.coinbase
eth.getBalance(eth.coinbase)
admin.nodeInfo.enode
admin.peers
admin.addPeer("enode://d90080e667e8e10a3171bf5096b661e32af7b52e6ca3253d9236604454f7e59065eac90f3a7e6129c6118e7f17337d84c64f94c087001f3c81c82036b36a6698@127.0.0.1:30304?discport=0")
admin.peers
eth.accounts

test.js
const fs = require('fs');
const Web3 = require('web3')
const RLP = require('rlp');
const utils = require('./utils')
const Trie = require('merkle-patricia-tree')
const { keccak, encode, decode, toBuffer} = require('eth-util-lite')

const rpcURL = 'http://127.0.0.1:8545' // RPC URL goes here
const web3 = new Web3(rpcURL);

const contractABI = JSON.parse(fs.readFileSync('contractABI.txt', 'utf-8'));

const contract = new web3.eth.Contract(contractABI, "0x3b4ec9e5e3f3cc80c7b8168066db041980a56157");
contract.methods.redeem("password").estimateGas().then(console.log).catch(console.log);

// var gasEstimate = web3.eth.estimateGas({
//     from: web3.eth.coinbase,
//     to: "0xc7278d22c0d1fecd9ea85a12d9b17815301eff45",
//     data: callData
// });

// var gasPrice = web3.eth.gasPrice;

// console.log('gas Price: ' + gasPrice);
// console.log('Estimated Transaction gas: ' + gasEstimate);

utils.js
const Web3 = require('web3');
const rpcURL = 'http://127.0.0.1:8545'; // Your RPC URL goes here
const web3 = new Web3(rpcURL);
const RLP = require('rlp');
const Trie = require('merkle-patricia-tree');
const { keccak, encode, decode, toBuffer} = require('eth-util-lite');
//const GetAndVerify = require('./helpers/getAndVerify');
const { GetAndVerify } = require('eth-proof'); // Check helpers/getAndVerify for code

let getAndVerify = new GetAndVerify(rpcURL);



module.exports = {
	verifyHeaderHash: function (blockHeader){
		function hashSerializer(hash){
			if(hash === 0) return '0x';
			return web3.utils.toHex(hash);
		}

		var arrFormat = [];
		arrFormat.push(blockHeader.parentHash);
		arrFormat.push(blockHeader.sha3Uncles);
		arrFormat.push(blockHeader.miner);
		arrFormat.push(blockHeader.stateRoot);
		arrFormat.push(blockHeader.transactionsRoot);
		arrFormat.push(blockHeader.receiptsRoot);
		arrFormat.push(blockHeader.logsBloom);
		arrFormat.push(blockHeader.difficulty);
		arrFormat.push(blockHeader.number);
		arrFormat.push(blockHeader.gasLimit);
		arrFormat.push(blockHeader.gasUsed);
		arrFormat.push(blockHeader.timestamp);
		arrFormat.push(blockHeader.extraData);
		arrFormat.push(blockHeader.mixHash);
		arrFormat.push(blockHeader.nonce);

		arrFormat = arrFormat.map(value => hashSerializer(value)); // Serialize hash
		arrFormat = '0x' + RLP.encode(arrFormat).toString('hex'); // RLP encode into hex
		return blockHeader.hash == web3.utils.keccak256(arrFormat);
	},

	verifyAccountStateRoot: async function (smartContractAddress, storageHash, nonce, codeHash, balance, proof, stateRoot){
		var check = true;
		// Verify that contract account address (merkleProof.address) depends on stateRoot (nonce, balance, storageHash, codeHash)
		// Nonce, balance, storageRoot, and codeHash are valid
		await Trie.verifyProof(stateRoot, keccak(smartContractAddress), proof, (err, value) => {
		   if (err || !value.equals(encode([parseInt(nonce),parseInt(balance),storageHash, codeHash])))
		      check = false;
		});

		return check;
	}, 

	verifyValidCode: async function(byteCode, codehash, originalContractCode, sendToAddress) {
		// NOTE replace with actual contract code
		let code = "pragmasolidity^0.4.11;contractTrent{addressowner;addresssender=0xBF20D8D746E53b6Bc1262CCa7c90276F6F0707d1;uintpublicvalue;//enumState{PUBLISHED,REDEEMED,REFUNDED}//Statepublicstate;bytes32publicconstantredeemHash=0x26ef0a97332ad048ac544d8dfacdd43e128b9816d599fdd0310960904fa18609;//keccak256hashof'redeem'keybytes32publicconstantrefundHash=0x4fd967966e6b0ea40f78dff297fed3b472763137dac6cf564d7263e918f425ef;//keccak256hashof'refund'keyconstructor()publicpayable{owner=msg.sender;value=msg.value;}functionredeem(stringkey)public{require(redeemHash==keccak256(abi.encodePacked(key)));//Sendsomemoneytootheraddresssender.transfer(value);//state=State.REDEEMED;}functionrefund(stringkey)public{require(redeemHash==keccak256(abi.encodePacked(key)));//Sendsomemoneytootheraddresssender.transfer(value);}functiondestroy()payablepublic{require(msg.sender==owner);selfdestruct(owner);}}"
		function replaceRange(s, start, end, substitute) {
		    return s.substring(0, start) + substitute + s.substring(end);
		}
		var check = true;

		// Replace sendToAddress at proper character count # 
		// and check if originalContractCode == sendToAddress
		check = originalContractCode.replace(/\s/g, '') == replaceRange(code, 63, 105, sendToAddress); 

		// Don't actually need below code assuming you can just read contract code directly
		// Check that codehash matches bytecode, so codehash is valid
		check =  web3.utils.keccak256(byteCode) == codehash;
		return check;
	}, 

	verifyStorage: async function(smartContractAddress, storageHash, storageProofs, fromAddress, contractFunds, blockHeaderHash){
		// TODO fix bug
		var check = true;

		// For every storageProof array index, check that it links to storageHash
		return Promise.all(storageProofs.map((proof) => {
			var check = true;
			// From https://github.com/zmitton/eth-proof/blob/master/getAndVerify.js
			// Check https://github.com/aragon/evm-storage-proofs for on-chain verification

			// Validate storage merkle proof checks out
			if(!getAndVerify.storageAgainstBlockHash(smartContractAddress, proof.key, blockHeaderHash)) check = false;
			
			// Verify msg.sender = fromAddress
			if(proof.key == 0 && proof.value != fromAddress) check = false;
			
			// Verify msg.val contains sufficient funds
			if(proof.key == 2 && proof.value != contractFunds) check = false;

			return check;
			/*
			//const path = web3.utils.soliditySha3({t: 'uint256', v: proof.key }).slice(2);
			console.log(web3.utils.padLeft(proof.key, 32))
			console.log(keccak(web3.utils.padLeft(proof.key, 32)))
			//console.log(path)
			//console.log(Buffer.from(path, 'hex'));
			return Trie.verifyProof(storageHash, keccak(web3.utils.padLeft(proof.key, 32)), proof.proof, (err, value) => {
			   if (err || !value.equals(proof.value)) { 
			   	console.log(err);
			   	return false;
			   }
			   else console.log(value)
			});*/
		})).then(result => result.every(check => check)).catch(console.log("Error when validating storage"));
	},


	/*verifyProof: function(rootHash, key, proof, cb) {
	  key = stringToNibbles(key)
	  var wantHash = ethUtil.toBuffer(rootHash)
	  for (var i = 0; i < proof.length; i++) {
	    var p = ethUtil.toBuffer(proof[i])
	    var hash = ethUtil.sha3(proof[i])
	    if (Buffer.compare(hash, wantHash)) {
	      return cb(new Error('Bad proof node ' + i + ': hash mismatch'))
	    }
	    var node = new TrieNode(ethUtil.rlp.decode(p))
	    var cld
	    if (node.type === 'branch') {
	      if (key.length === 0) {
	        if (i !== proof.length - 1) {
	          return cb(new Error('Additional nodes at end of proof (branch)'))
	        }
	        return cb(null, node.value)
	      }
	      cld = node.raw[key[0]]
	      key = key.slice(1)
	      if (cld.length === 2) {
	        var embeddedNode = new TrieNode(cld)
	        if (i !== proof.length - 1) {
	          return cb(new Error('Additional nodes at end of proof (embeddedNode)'))
	        }

	        if (matchingNibbleLength(embeddedNode.key, key) !== embeddedNode.key.length) {
	          return cb(new Error('Key length does not match with the proof one (embeddedNode)'))
	        }
	        key = key.slice(embeddedNode.key.length)
	        if (key.length !== 0) {
	          return cb(new Error('Key does not match with the proof one (embeddedNode)'))
	        }
	        return cb(null, embeddedNode.value)
	      } else {
	        wantHash = cld
	      }
	    } else if ((node.type === 'extention') || (node.type === 'leaf')) {
	      if (matchingNibbleLength(node.key, key) !== node.key.length) {
	        return cb(new Error('Key does not match with the proof one (extention|leaf)'))
	      }
	      cld = node.value
	      key = key.slice(node.key.length)
	      if (key.length === 0 || (cld.length === 17 && key.length === 1)) {
	        // The value is in an embedded branch. Extract it.
	        if (cld.length === 17) {
	          cld = cld[key[0]][1]
	          key = key.slice(1)
	        }
	        if (i !== proof.length - 1) {
	          return cb(new Error('Additional nodes at end of proof (extention|leaf)'))
	        }
	        return cb(null, cld)
	      } else {
	        wantHash = cld
	      }
	    } else {
	      return cb(new Error('Invalid node type'))
	    }
	  }
	}*/
}

queryBlockchain.js
const Web3 = require('web3')
const fs = require('fs')
const rpcURL = 'http://127.0.0.1:8545' // Your RPC URL goes here
const web3 = new Web3(rpcURL);

const debug = false;
const smartContractAddress = "0xfb32abae531d7e4cc4d71507de558ee3a24a74ff"; // Change contract address here


function retrieve(){ // Retrieve data and write to file to be used for trent.js 

	// Retrieve block headers and write to blockHeaders.json
	retrieveBlockHeaders();

	// Retrieve bytecode of contract and write to contractCode.txt
	retrieveContractCode(smartContractAddress);

	// Retrieve storage / contract account merkle proof 
	// and the data itself (sender/recipient address (msg.sender), msg.value).
	// Then write to merkleDataProof.json
	retrieveProof(smartContractAddress);
}

retrieve();

async function retrieveBlockHeaders(){
	var blocks = []; 
	var isNextBlock = true;
	var i = 0;

	while(isNextBlock){
		await web3.eth.getBlock(i).then((result) => { 
			if(debug) console.log("Inside eth.getBlock retrieveBlock");
			if(result != null) blocks.push(result);
			else isNextBlock = false;
		});
		i++;
	}

	if(debug) console.log(`${blocks.length} blocks`);
	if(debug) console.log(blocks);

	fs.writeFile('data/blockHeaders.json', JSON.stringify(blocks), (err) => {
		if(err) throw err;
	});
}

function retrieveContractCode(address){
	// Contains raw code
	web3.eth.getCode(address).then((result) => {
		fs.writeFile('data/byteCode.txt', result, (err) => { if(err) throw err});
	});
}

async function retrieveProof(address){
	// Contains necessary information for merkle proof
	// Retrieves proof for contract account containing code, sender/recipient addresses (msg.sender), msg.value, codehash, etc.
	web3.eth.getProof(address, ["0", "2"], "latest").then((result) => {
		fs.writeFile('data/merkleDataProof.json', JSON.stringify(result), (err) => {
			if(err) throw err;
		});
	});
}

//console.log(web3.utils.toChecksumAddress("0x123f681646d4a755815f9cb19e1acc8565a0c2ac"))
// Try both and do string comparison on bytecode to figure out ways to check
//0x8ef93287e177afe79472f70d1b2c9d806844c931 
// web3.eth.getCode("0x46465f376d3034d5f6aa1ee5a43b6ce572ebb153").then((result) => {
// 	console.log(result)
// 	});


/*
Address querying
const address = '0xccb06e380ae6805c7e6c77be24e215bb3442735a' // Your account address goes here
web3.eth.getBalance(address, (err, wei) => {
  balance = web3.utils.fromWei(wei, 'ether');
  console.log(balance)
	web3.utils.isAddress('0xc1912fee45d61c87cc5ea59dae31190fffff232d');

})
console.log(web3.eth.getTransactionCount());


Used for debugging storage
async function helper_printStorage(address){
	for(var i = 0; i < 10; i++){
		web3.eth.getStorageAt(smartContractAddress, i).then((value) => {
			if(value != 0) console.log(value)
		})
	}
}

function playground(){
	web3.eth.getCode(contractAddress).then(console.log);
	web3.eth.getTransactionCount(smartContractAddress).then(console.log);
	web3.eth.getStorageAt(smartContractAddress).then((result) => {
		console.log(result);
		console.log(web3.utils.hexToAscii(result));
	});
	web3.eth.getProof(smartContractAddress, ["0", "2"]).then((result) => console.log(JSON.stringify(result)));
}

*/



helpers/getProof.js
const { encode } = require('eth-util-lite')
const { promisfy } = require('promisfy')

const Tree = require('merkle-patricia-tree')

const Rpc  = require('isomorphic-rpc')

const { Header, Proof, Receipt, Transaction } = require('eth-object')


module.exports = class GetProof{
  constructor(rpcProvider = "https://mainnet.infura.io"){
    this.rpc = new Rpc(rpcProvider)
    this.eth_getProof = this.rpc.eth_getProof
  }

  async transactionProof(txHash){
    let targetTx = await this.rpc.eth_getTransactionByHash(txHash)
    if(!targetTx){ throw new Error("Tx not found. Use archive node")}

    let rpcBlock = await this.rpc.eth_getBlockByHash(targetTx.blockHash, true)

    let tree = new Tree();

    await Promise.all(rpcBlock.transactions.map((siblingTx, index) => {
      let siblingPath = encode(index)
      let serializedSiblingTx = Transaction.fromRpc(siblingTx).serialize()
      return promisfy(tree.put, tree)(siblingPath, serializedSiblingTx) 
    }))

    let [_,__,stack] = await promisfy(tree.findPath, tree)(encode(targetTx.transactionIndex))

    return {
      header:  Header.fromRpc(rpcBlock),
      txProof:  Proof.fromStack(stack),
      txIndex: targetTx.transactionIndex,
    }
  }
  async receiptProof(txHash){
    let targetReceipt = await this.rpc.eth_getTransactionReceipt(txHash)
    if(!targetReceipt){ throw new Error("txhash/targetReceipt not found. (use Archive node)")}

    let rpcBlock = await this.rpc.eth_getBlockByHash(targetReceipt.blockHash, false)

    let receipts = await Promise.all(rpcBlock.transactions.map((siblingTxHash) => {
      return this.rpc.eth_getTransactionReceipt(siblingTxHash)
    }))

    let tree = new Tree();

    await Promise.all(receipts.map((siblingReceipt, index) => {
      let siblingPath = encode(index)
      let serializedReceipt = Receipt.fromRpc(siblingReceipt).serialize()
      return promisfy(tree.put, tree)(siblingPath, serializedReceipt)
    }))

    let [_,__,stack] = await promisfy(tree.findPath, tree)(encode(targetReceipt.transactionIndex))

    return {
      header:  Header.fromRpc(rpcBlock),
      receiptProof:  Proof.fromStack(stack),
      txIndex: targetReceipt.transactionIndex,
    }
  }
  async accountProof(address, blockHash = null){
    let rpcBlock, rpcProof
    if(blockHash){
      rpcBlock = await this.rpc.eth_getBlockByHash(blockHash, false)
    }else{
      rpcBlock = await this.rpc.eth_getBlockByNumber('latest', false)
    }
    rpcProof = await this.eth_getProof(address, [], rpcBlock.number)

    return {
      header: Header.fromRpc(rpcBlock),
      accountProof: Proof.fromRpc(rpcProof.accountProof),
    }
  }
  async storageProof(address, storageAddress, blockHash = null){
    let rpcBlock, rpcProof
    if(blockHash){
      rpcBlock = await this.rpc.eth_getBlockByHash(blockHash, false)
    }else{
      rpcBlock = await this.rpc.eth_getBlockByNumber('latest', false)
    }
    rpcProof = await this.eth_getProof(address, [storageAddress], rpcBlock.number)

    return {
      header: Header.fromRpc(rpcBlock),
      accountProof: Proof.fromRpc(rpcProof.accountProof),
      storageProof: Proof.fromRpc(rpcProof.storageProof[0].proof),
    }
  }
}

helpers/verifyProof.js
const { keccak, encode, decode, toBuffer, toWord } = require('eth-util-lite')

const Tree = require('merkle-patricia-tree')

const { Account, Receipt, Transaction } = require('eth-object')

const ACCOUNTS_ROOT_INDEX = 3 // within header
const TXS_ROOT_INDEX      = 4 // within header
const RECEIPTS_ROOT_INDEX = 5 // within header

const STORAGE_ROOT_INDEX  = 2 // within account
const SET_OF_LOGS_INDEX   = 3 // within receipt

class Verify{

  static chainContainsBlockHash(_chain, _blockHash){ throw new Error("Feature not yet available") }

  static getRootFromProof(proof){ return keccak(encode(proof[0])) }

  static accountContainsStorageRoot(account){
    return account[STORAGE_ROOT_INDEX]
  }
  static getBlockHashFromHeader(header){
    return keccak(encode(header))
  }
  static getElemFromHeaderAt(header, indexOfRoot){
    return header[indexOfRoot]
  }
  static getStateRootFromHeader(header){
    return this.getElemFromHeaderAt(header, ACCOUNTS_ROOT_INDEX)
  }
  static getTxsRootFromHeader(header){
    return this.getElemFromHeaderAt(header, TXS_ROOT_INDEX)
  }
  static getReceiptsRootFromHeader(header){
    return this.getElemFromHeaderAt(header, RECEIPTS_ROOT_INDEX)
  }
  static receiptContainsLogAt(receipt, indexOfLog){
    return receipt[SET_OF_LOGS_INDEX][indexOfLog]
  }

  static async getStorageFromStorageProofAt(proof, position){
    let fromProof = await this.proofContainsValueAt(proof, keccak(toWord(position)))
    if(!fromProof || fromProof.equals(toBuffer())){
      return toBuffer() // null returned as <buffer >
    }else{
      return decode(fromProof)
    }
  }
  static async getAccountFromProofAt(proof, address){
    let accountBuffer = await this.proofContainsValueAt(proof, keccak(address))
    return Account.fromBuffer(accountBuffer) // null returned as Account.NULL
  }
  static async getTxFromTxProofAt(proof, indexOfTx){
    let txBuffer = await this.proofContainsValueAt(proof, encode(indexOfTx))
    return Transaction.fromBuffer(txBuffer)
  }
  static async getReceiptFromReceiptProofAt(proof, indexOfTx){
    let receiptBuffer = await this.proofContainsValueAt(proof, encode(indexOfTx))
    return Receipt.fromBuffer(receiptBuffer)
  }

  static async proofContainsValueAt(proof, path){
    return new Promise((accept, reject) => {
      let encodedProof = []
      for (let i = 0; i < proof.length; i++) {
        encodedProof.push(encode(proof[i]))
      }

      Tree.verifyProof(toBuffer(this.getRootFromProof(proof)) , path, encodedProof, (e,r)=>{
        if(e){
          return reject(e)
        }else{
          return accept(r)
        }
      })
    })
  }
}

module.exports = Verify

helpers/getAndVerify.js
const GetProof = require('./getProof')
const VerifyProof = require('./verifyProof')
const { toBuffer } = require('eth-util-lite')

class GetAndVerify{
  
  constructor(rpcProvider = "https://mainnet.infura.io"){
    this.get = new GetProof(rpcProvider)
  }

  async txAgainstBlockHash(txHash, trustedBlockHash){
    let resp = await this.get.transactionProof(txHash)
    let blockHashFromHeader = VerifyProof.getBlockHashFromHeader(resp.header)
    if(!toBuffer(trustedBlockHash).equals(blockHashFromHeader)) throw new Error('BlockHash mismatch')
    let txRootFromHeader = VerifyProof.getTxsRootFromHeader(resp.header)
    let txRootFromProof = VerifyProof.getRootFromProof(resp.txProof)
    if(!txRootFromHeader.equals(txRootFromProof)) throw new Error('TxRoot mismatch')
    return VerifyProof.getTxFromTxProofAt(resp.txProof, resp.txIndex)
  }
  async receiptAgainstBlockHash(txHash, trustedBlockHash){
    let resp = await this.get.receiptProof(txHash)
    let blockHashFromHeader = VerifyProof.getBlockHashFromHeader(resp.header)
    if(!toBuffer(trustedBlockHash).equals(blockHashFromHeader)) throw new Error('BlockHash mismatch')
    let receiptsRoot = VerifyProof.getReceiptsRootFromHeader(resp.header)
    let receiptsRootFromProof = VerifyProof.getRootFromProof(resp.receiptProof)
    if(!receiptsRoot.equals(receiptsRootFromProof)) throw new Error('ReceiptsRoot mismatch')
    return VerifyProof.getReceiptFromReceiptProofAt(resp.receiptProof, resp.txIndex)
  }
  async accountAgainstBlockHash(accountAddress, trustedBlockHash){
    let resp = await this.get.accountProof(accountAddress, trustedBlockHash)
    let blockHashFromHeader = VerifyProof.getBlockHashFromHeader(resp.header)
    if(!toBuffer(trustedBlockHash).equals(blockHashFromHeader)) throw new Error('BlockHash mismatch')
    let stateRoot = VerifyProof.getStateRootFromHeader(resp.header)
    let stateRootFromProof = VerifyProof.getRootFromProof(resp.accountProof)
    if(!stateRoot.equals(stateRootFromProof)) throw new Error('StateRoot mismatch')
    return VerifyProof.getAccountFromProofAt(resp.accountProof, accountAddress)
  }
  async storageAgainstBlockHash(accountAddress, position, trustedBlockHash){
    let resp = await this.get.storageProof(accountAddress, position, trustedBlockHash)
    let blockHashFromHeader = VerifyProof.getBlockHashFromHeader(resp.header)
    if(!toBuffer(trustedBlockHash).equals(blockHashFromHeader)) throw new Error('BlockHash mismatch')
    let stateRoot = VerifyProof.getStateRootFromHeader(resp.header)
    let stateRootFromProof = VerifyProof.getRootFromProof(resp.accountProof)
    if(!stateRoot.equals(stateRootFromProof)) throw new Error('StateRoot mismatch')
    let account = await VerifyProof.getAccountFromProofAt(resp.accountProof, accountAddress)
    let storageRoot = VerifyProof.accountContainsStorageRoot(account)
    let storageRootFromProof = VerifyProof.getRootFromProof(resp.storageProof)
    if(!storageRoot.equals(storageRootFromProof)) throw new Error('StorageRoot mismatch')
    return VerifyProof.getStorageFromStorageProofAt(resp.storageProof, position)
  }

  async _logAgainstBlockHash(txHash, indexOfLog, trustedBlockHash){
    // untested as of yet
    let resp = await this.get.receiptProof(txHash)
    let blockHashFromHeader = VerifyProof.getBlockHashFromHeader(resp.header)
    if(!toBuffer(trustedBlockHash).equals(blockHashFromHeader)) throw new Error('BlockHash mismatch')
    let receiptsRoot = VerifyProof.getReceiptsRootFromHeader(resp.header)
    let receiptsRootFromProof = VerifyProof.getRootFromProof(resp.receiptProof)
    if(!receiptsRoot.equals(receiptsRootFromProof)) throw new Error('ReceiptsRoot mismatch')
    let receipt = await VerifyProof.getReceiptFromReceiptProofAt(resp.receiptProof, resp.txIndex)
    return VerifyProof.receiptContainsLogAt(receipt, indexOfLog)
  }
}

module.exports = GetAndVerify

trent.js
const fs = require('fs');
const Web3 = require('web3')
const RLP = require('rlp');
const utils = require('./utils')
const Trie = require('merkle-patricia-tree')
const { keccak, encode, decode, toBuffer} = require('eth-util-lite')

const rpcURL = 'http://127.0.0.1:8545' // RPC URL goes here
const web3 = new Web3(rpcURL);


attemptRedeem();

function attemptRedeem(){
	// Calls redeem 
	console.log("Issuing redeem signature") 
	verify().then(result => { 
		if(result) console.log(web3.eth.abi.encodeParameter('bytes32', keccak('redeem')))
		else console.log("Failed to redeem");
	});
}

function attemptRefund(){
	console.log("Issuing refund signature") 
	console.log(web3.eth.abi.encodeParameter('bytes32', keccak('refund')));
}

// Give signature if everything is given correctly
function verify(){
	const byteCode = fs.readFileSync('data/byteCode.txt', 'utf-8');
	const blockHeaders = JSON.parse(fs.readFileSync('data/blockHeaders.json', 'utf-8'));
	const merkleProof = JSON.parse(fs.readFileSync('data/merkleDataProof.json', 'utf-8'));
	const smartContractCode = fs.readFileSync('data/smartContractCode.txt', 'utf-8');
	const fromAddress = 0xCCb06E380ae6805C7e6c77bE24E215BB3442735A; // NOTE Modify this with correct address
	const sendToAddress = 0xBF20D8D746E53b6Bc1262CCa7c90276F6F0707d1; // NOTE Modify this with correct address

	const smartContractAddress1 = "0xfb32abae531d7e4cc4d71507de558ee3a24a74ff"; // NOTE Change contract address here
	const smartContractAddress2 = ""; // NOTE Change contract address here

	const contractFunds = 1000;  // NOTE change this value depending on agreed upon funds
	const d = blockHeaders.length;

	// if all pass, return signature
	return Promise.all([
		verifyBlockHeaders(blockHeaders, d), 
		verifyMerkleProof(blockHeaders[blockHeaders.length-1], merkleProof, byteCode, fromAddress, sendToAddress, smartContractCode, smartContractAddress1, contractFunds)
		/*verifyMerkleProof(blockHeaders[blockHeaders.length-1], merkleProof, byteCode, sendToAddress, fromAddress, smartContractCode, smartContractAddress2, contractFunds)*/])
	.then((result) => {
		console.log("Final result: " + result)
		if(result.every(a => a)) return true;
	}).catch(console.log);
}

async function verifyBlockHeaders(blockHeaders, d){ 
	var parentHash = blockHeaders[0].parentHash;
	return blockHeaders.every(function(block){
		var check = true;

		// Check if it's at least d blocks deep
		if(blockHeaders.length < d) check = false;

		// Compute hash of blockheader data and see if it equals to hash
		if(!utils.verifyHeaderHash(block)) check = false;

		// Verify this block contains previous hash 
		if(block.parentHash !== parentHash) check = false;
		
		// Verify difficulty is within some range. 
		// NOTE -- replace 130000 difficulty with correct difficulty (most recent ethereum difficulty ~ 2,000,000,000,000,000)
		// OR replace with calculating for both BTC (changes every epoch) and ETH (calculation in white paper)
		if(block.number != 0) check = block.difficulty > 130000; 

		// Verify valid PoW (nonce correctly calculated)
		// NOTE rewrite my own PoW verification, look at internals
		web3.eth.submitWork(block.nonce, block.hash, block.mixHash)
		.then((result) => check = result)
		.catch((err) => {if(err) check = false}); // when node is shutdown

		parentHash = block.hash;
		return check;
	});
}

async function verifyMerkleProof(latestBlockHeader, mp, byteCode, fromAddress, sendToAddress, originalContractCode, smartContractAddress, contractFunds){
	//  Look at internals at https://github.com/zmitton/eth-proof
	// and this https://github.com/ethereumjs/merkle-patricia-tree
	// For Ethereum on chain merkle proof verification - https://github.com/lorenzb/proveth

	var check = true;
	// Valid merkle proof that contract account address (merkleProof.address) depends on stateRoot
	// and is dependent on provided storageroot, codeHash, nonce, balance
	accountResult = utils.verifyAccountStateRoot(smartContractAddress, mp.storageHash, mp.nonce, mp.codeHash, mp.balance, mp.accountProof, latestBlockHeader.stateRoot)

	// Check if code is correct + bytecode is correct codeHash (don't need this if you assume you have contract code)
	validCodeResult = utils.verifyValidCode(byteCode, mp.codeHash, originalContractCode, sendToAddress);

	// Check storage data is correct and dependent on storageRoot 
	// Validate msg.sender, recipient address, and msg.val
	storageResult = utils.verifyStorage(smartContractAddress, mp.storageHash, mp.storageProof, fromAddress, contractFunds, latestBlockHeader.hash);

	return Promise.all([accountResult, storageResult, validCodeResult]).then((values) => {
		return values.every((result) => result);
	}).catch(console.log);
}


